gsoc-2025/semantic-routines.md

```diff
@@ -0,0 +1,29 @@
+---
+parent: DLang GSoC 2025 Project Ideas
+nav_order: 1
+---
+
+# Separate Semantic Routines From AST Nodes
+
+**Mentor:** [Razvan Nitu](razvan.nitu1305@gmail.com)
+
+| Spec | Details |
+|-|-|
+| **Difficulty Level** | easy-medium |
+| **Project Duration** | 200 hours |
+| **Number of Contributors** | 1 |
+| **Prerequisites** | Familiarity with compiler organization, visitor pattern, object oriented programming |
+
+## Description
+
+In the DMD compiler codebase, AST nodes are defined as classes within various files. The ideal structure for these nodes is to have minimal fields and methods focused solely on field queries. However, the current state of the DMD frontend deviates from this ideal. AST nodes are laden with numerous methods that either perform or are dependent on semantic analysis. Furthermore, many AST node files contain free functions related to semantic analysis. Our objective is to decouple AST nodes from these functions, both directly and indirectly.
+
+This project helps advance the development of the compiler library by creating a clear separation between compilation phases.
+
+This project is ideal for someone that has no prior experience with compilers but wants to understand how a compiler works by doing valuable work.
+
+## Resources
+
+- [Compiler codebase](https://github.com/dlang/dmd/tree/master/compiler/src/dmd)
+- [List of files that need to have semantic separated out of them](https://github.com/orgs/dlang/projects/41/views/1)
+- [How to start](https://github.com/dlang/dmd/blob/master/CONTRIBUTING.md#refactoring-the-dmd-ast)
```

Comment:
- Add this guide about building the compiler: https://wiki.dlang.org/Building_under_Posix#Building_DMD
- Add previous PRs on this subject as examples